flink-connector/docs/content/docs/connectors/datastream/pubsub.md

```diff
@@ -0,0 +1,158 @@
+Google Cloud Pub/Sub has no notion of subscription partitions or splits, a
+message can be received by any reader. Google Cloud Pub/Sub automatically load
+balances message delivery across readers.
+
+### Deserialization Schema
+
+`PubSubDeserializationSchema<T>` is required to define how
+[`PubsubMessage`](https://cloud.google.com/pubsub/docs/reference/rest/v1/PubsubMessage)
+is deserialized into an output that is suitable for processing.
+
+For convenience, `PubSubDeserializationSchema.dataOnly` can be used if only the
+field `PubsubMessage.data` is required for processing.
+
+### Boundedness
+
+Pub/Sub source streams unbounded data, only stopping when a Flink job stops or
+fails.
+
+### Checkpointing
+
+Checkpointing is required to use Pub/Sub source. When a checkpoint completes,
+all messages delivered before the last successful checkpoint are acknowledged to
+Google Cloud Pub/Sub. In case of failure, messages delivered after the last
+successful checkpoint are unacknowledged and will automatically be redelivered.
+Note that there is no message delivery state stored in checkpoints, so retained
+checkpoints are not necessary to resume using Pub/Sub source.
+
+### Flow Control
+
+Each Pub/Sub source subtask manages a `StreamingPull` connection to Google Cloud
+Pub/Sub. That is, a Pub/Sub source with parallelism of 20 manages 20 separate
+`StreamingPull` connections to Google Cloud Pub/Sub. The flow control settings
+described in this section are applied to **individual** connections.
+
+Several Pub/Sub source options configure flow control. These options are
+illustrated below:
+
+```java
+PubSubSource.<OutputT>builder()
+    // Allow up to 10,000 message deliveries per checkpoint interval.
+    .setMaxOutstandingMessagesCount(10_000L)
+    // Allow up to 1000 MB in cumulatitive message size per checkpoint interval.
+    .setMaxOutstandingMessagesBytes(1000L * 1024L * 1024L)  // 1000 MB
+```
+
+Google Cloud Pub/Sub servers pause message delivery when a flow control setting
+is exceeded. A message is considered outstanding from when Google
+Cloud Pub/Sub delivers it until the subscriber acknowledges it. Since
+outstanding messages are acknowledged when a checkpoint completes, flow control
+limits for outstanding messages are effectively per-checkpoint interval limits.
+
+### Message Leasing
+
+Pub/Sub source automatically extends the acknowledge deadline of messages. This
+means a checkpointing interval can be longer than the acknowledge deadline
+without causing messages redelivery. Note that acknowledge deadlines can be
+extended to at most 1h.
+
+### Performance Considerations
+
+- Infrequent checkpointing can cause performance issues, including subscription
+  backlog growth and increased rate of duplicate message delivery.
+- `StreamingPull` connections have a 10 MB/s throughput limit. See
+  [Pub/Sub quotas and limits](https://cloud.google.com/pubsub/quotas) for all
+  restrictions.
+
+<!-- TODO(matt-kwong) Add threading details. -->
```

Comment:
is this gonna be in a follow-up PR?

Reply:
Yes, I'm planning to expose additional builder options that influence threading.